components/raftstore-v2/src/operation/command/admin/compact_log.rs

```diff
@@ -0,0 +1,144 @@
+// Copyright 2022 TiKV Project Authors. Licensed under Apache-2.0.
+
+//! `CompactLog` command.
+
+use engine_traits::{KvEngine, RaftEngine};
+use kvproto::raft_cmdpb::{AdminRequest, AdminResponse, RaftCmdRequest};
+use protobuf::Message;
+use raftstore::Result;
+use slog::{debug, error, info};
+use tikv_util::box_err;
+
+use crate::{
+    batch::StoreContext,
+    fsm::{ApplyResReporter, PeerFsmDelegate},
+    operation::AdminCmdResult,
+    raft::{write_initial_states, Apply, Peer, Storage},
+    router::{ApplyRes, PeerMsg, StoreMsg},
+    worker::RaftLogGcTask,
+};
+
+#[derive(Debug)]
+pub struct CompactLogResult {
+    compact_index: u64,
+    compact_term: u64,
+}
+
+impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
+    pub fn propose_compact_log<T>(
+        &mut self,
+        store_ctx: &mut StoreContext<EK, ER, T>,
+        req: RaftCmdRequest,
+    ) -> Result<u64> {
+        let compact_log = req.get_admin_request().get_compact_log();
+        // TODO: add unit tests to cover all the message integrity checks.
+        if compact_log.get_compact_term() == 0 {
+            info!(
+                self.logger,
+                "compact term missing, skip";
+                "region_id" => self.region_id(),
+                "peer_id" => self.peer_id(),
+                "command" => ?compact_log
+            );
+            // old format compact log command, safe to ignore.
+            return Err(box_err!(
+                "command format is outdated, please upgrade leader"
+            ));
+        }
+
+        let data = req.write_to_bytes().unwrap();
+        self.propose_with_ctx(store_ctx, data, vec![])
+    }
+
+    pub fn on_ready_compact_log<T>(
```

Comment:
I have no idea if it's correct to move the v1 check in `exec_compact_log` to this v2 function.

Reply:
Because I seem to find access to apply_state in `Apply`, I'm not sure if it's supposed to be added.